Show counts from fusiontables

RobinIsTheBird opened the pull request:
Get counts from fusiontables and show them.

This is done with a minimum of third party scripts for the sake of build simplicity.
Therefore, dynamically add script tags, rather than using a JS module package
to get google maps with a production or development api key and use the right rawkit cdn for the fusiontables.
Also directly manipulate DOM to show counts.

Also create a `.gitignore` to ignore editor temp files.

RobinIsTheBird committed:
Show counts from fusiontables

Get counts from fusiontables and show them.

This is done with a minimum of third party scripts for the sake
of build simplicity.
Therefore,
dynamically add script tags, rather than using a JS module package
to get google maps with a production or development api key,
and use the right rawkit cdn for the fusiontables.
Also directly manipulate DOM to show counts.

Also create a .gitignore, ignore editor temp files

RobinIsTheBird commented on fdpa-resolutions.html:
```diff
@@ -2,79 +2,21 @@
 <html>
   <head>
   <title>Fair Districts PA - Local Resolutions</title>
+  <meta name="referrer" content="origin">
```
Necessary for the rawgit references below.

RobinIsTheBird replied:
Seems like an HTML comment is needed here. Something like
```
<!-- Necessary for rawgit and google fusiontables cross origin requests -->
```
You agree, @amyprime?

amyprime replied:
Yeah, and adding the fusiontables to the comment makes it even better.

RobinIsTheBird commented on fdpa-resolutions.html:
```diff
@@ -2,79 +2,21 @@
   <style>
     #map-canvas { width:1100px; height:650px; }
-    .layer-wizard-search-label { font-family: sans-serif };
+    .layer-wizard-search-label { font-family: sans-serif }
```
Semicolon after a style rule is incorrect syntax. Had to remove it to add another rule.

RobinIsTheBird commented on fdpa-resolutions.html:
```diff
@@ -86,5 +28,159 @@ <h2>Fair Districts PA - Local Resolutions</h2>
         &nbsp; | &nbsp; Click a county or municipality for more info &nbsp; | &nbsp;
         <a href="https://tinytake.com/how-to-take-a-screenshot/">Help with screen shots</a></p>
     <div id="map-canvas"></div>
+    <script src="https://www.promisejs.org/polyfills/promise-7.0.4.min.js">
```
Promises are the best way to join multiple async requests.

RobinIsTheBird commented on fdpa-resolutions.html:
```diff
@@ -86,5 +28,159 @@ <h2>Fair Districts PA - Local Resolutions</h2>
+      var map;
+      var layer_0;
+      var layer_1;
+      function initialize() {
```
initialize() is basically identical to the original JavaScript. It now gets called after dynamically injecting the google maps script with either the production or the development key.

RobinIsTheBird commented on fdpa-resolutions.html:
```diff
@@ -86,5 +28,159 @@ <h2>Fair Districts PA - Local Resolutions</h2>
+          templateId: 5
+        });
+      }
+      (function() {
```
Use an IIFE (immediately-invoked function expression) to avoid any further pollution of the global namespace.

RobinIsTheBird commented on fdpa-resolutions.html:
```diff
@@ -86,5 +28,159 @@ <h2>Fair Districts PA - Local Resolutions</h2>
+          fallbackCount: 'fall-back-count',
+          currentCount: 'current-count'
+        };
+        var windowPromise =
```
All this promise stuff is obvious to me, but if it isn't clear to the reviewer, it may be worth adding comments.

amyprime replied:
I don't follow it 100%, but it looks like I could if I tried harder, so I'm not sure comments are necessary.  And I think I'd mostly be spending that time thinking about the async nature than the promise syntax.  Maybe it would be good to just turn your note "Promises are the best way to join multiple async requests" above into a real comment in the code.

amyprime replied:
Okay, the new comments are super awesome.

RobinIsTheBird reviewed (commented):
Calling the reviewer's attention to things that might require a bit of explanation.

amyprime commented on fdpa-resolutions.html:
```diff
@@ -86,5 +28,159 @@ <h2>Fair Districts PA - Local Resolutions</h2>
+        new Promise(function(scriptAdded){
+          addScript(isProduction() ?
+            'https://cdn.rawgit.com/RobinIsTheBird/fdpa-fusiontables/master/distribution/fusionApi_es5.js' :
+            'https://rawgit.com/RobinIsTheBird/fdpa-fusiontables/master/distribution/fusionApi_es5.js',
```
I'm not seeing any difference between the two fusionApi_es5.js files (though I didn't fully diff them).  Why the switch?  This may be worth adding a comment.

RobinIsTheBird replied:
Sounds like I need to add a comment here. If you visit https://rawgit.com/, you'll see that it converts a github URL into not one but two rawgit URLs, one to use in production, the other to use in development. That's what this switch is all about. They both fetch the exact same .js file from github.

amyprime replied:
Ah, yup, that site explains it nicely.  A quick comment pointing people there would be helpful.

RobinIsTheBird replied:
How about something like:
```
/*
 * See rawgit.com for an explanation of why two URLs are necessary here.
 */

amyprime commented on fdpa-resolutions.html:
```diff
@@ -86,5 +28,159 @@ <h2>Fair Districts PA - Local Resolutions</h2>
+              fdpaResolutions.getResolutionsCount({
+                status: statusSought,
+                fusiontable: '10Uc_t_dBYUV_K_j6HdCMSgLXGs94bXvunQBGDOjF',
+                googleApiKey: getKey(),
```
I'm not positive I'm doing it right, but I'm trying to test with my original API key, and I can see the map, but for the counts I get: "The request did not specify any referer. Please ensure that the client is sending referer or use the API Console to remove the referer restrictions."

RobinIsTheBird replied:
I sent you email with a long explanation. Cutting to the quick here, I took the liberty to correct your API key restrictions and now it works.

amyprime replied:
Hooray!

amyprime commented on fdpa-resolutions.html:
```diff
@@ -2,79 +2,21 @@
+    <p id="current-count" class="hidden">To date,
+        <span data-status="passed"></span> counties and municipalities across Pennsylvania have adopted
+        a resolution in support of an independent citizens' redistricting commission.
+        <span data-status="inWork"></span> more are in progress.</p>
```
So here's a thing I forgot to explain... the yellow dots on the map are actually "In work" and "Materials sent" together.  It would be best for this number to match the number of yellow dots.  It looks like your API is set up to allow both queries, so can you just do a quick sum in the page here for injecting into the span?

RobinIsTheBird replied:
I'll change the fdpa-resolutions getResolutionsCount to accept an array in the status option, and it will send an OR in the sql query. That's how to get absolutely consistent numbers. Thanks for the clarification.

amyprime commented on fdpa-resolutions.html:
```diff
@@ -2,79 +2,21 @@
     <h2>Fair Districts PA - Local Resolutions</h2>
-    <p>To date, more than 130 counties and municipalities across Pennsylvania have adopted a
+    <p id="fall-back-count">To date, more than 198 counties and municipalities across Pennsylvania have adopted a
         resolution in support of an indpendent citizens' redistricting commission.</p>
```
Note: There's a merge conflict now because I fixed the spelling of "indpendent".  Should be trivial to resolve.

RobinIsTheBird replied:
Okay. With PRs, it is procedurally simpler to tell the author of the PR and let the author make the change, but given that neither of us is doing this f/t, maybe that stretches out the time period too much.

RobinIsTheBird replied:
I see, you made the fix in `master`. Merging it is no problem.

RobinIsTheBird committed:
Review remarks addressed

Changes to include materials sent in the in progress count depend
upon a change to the fusionApi script that are currently in a
branch awaiting review. Therefore, the rawgit URIs now refer to
the branch rather than master for the sake of testing.

When the fusionApi script is merged to its master,
we have to change the rawgit URIs.

Also sprinkled with comments.

RobinIsTheBird commented:
Amy,

I modified the fdpa-fusiontables repo in a branch, invited you to
collaborate, and created a PR.

Please accept the invitation, add yourself as a reviewer on the PR, and do
the review, before you check out my updates to the github.io PR. I need to
merge the fdpa-fusiontables branch before we can merge the github.io PR,
because the rawgit calls in fdpa-resolutions.html presently refer to the
fdpa-fusiontables branch. They need to be put back to master before merge.

Thank you ever so much for collaborating with me!

Robin, like the bird
cell: 650-906-1784
home: 610-328-0510

On Wed, Feb 28, 2018 at 8:15 AM, amyprime <notifications@github.com> wrote:

> *@amyprime* commented on this pull request.
> ------------------------------
>
> In fdpa-resolutions.html
> <https://github.com/amyprime/amyprime.github.io/pull/1#discussion_r171240381>
> :
>
> > +      }
> +      (function() {
> +        var apiKey = {
> +          amyprime: 'AIzaSyBxdPyKNXkcFbFPs5I_e-xqGCUPGVuyabs',
> +          devel: 'AIzaSyBz3CQw1aK0S9UHwxKq-YZtIeqdcOby-Jc'
> +        };
> +        var locationUrl = new URL(window.location.href);
> +
> +        addGoogleMaps();
> +
> +        var statuses = ['passed', 'inWork'];
> +        var countElIds = {
> +          fallbackCount: 'fall-back-count',
> +          currentCount: 'current-count'
> +        };
> +        var windowPromise =
>
> I don't follow it 100%, but it looks like I could if I tried harder, so
> I'm not sure comments are necessary. And I think I'd mostly be spending
> that time thinking about the async nature than the promise syntax. Maybe it
> would be good to just turn your note "Promises are the best way to join
> multiple async requests" above into a real comment in the code.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/amyprime/amyprime.github.io/pull/1#discussion_r171240381>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABZ678uMnsWmkZUu81PfyywTRWvZE0gLks5tZVF7gaJpZM4SUPJX>
> .
>

RobinIsTheBird committed:
Restore fusionApi_es5.js to master

The rawgit cdn fetch of fusionApi_es5.js had been changed to point
at a branch which contained support for multiple statuses.
That branch has been merged and deleted, so point back to master.

RobinIsTheBird committed:
Merge branch 'master' into show-resolutions-count
